lib/web/app/handler.go

```diff
@@ -232,8 +232,31 @@ func (h *Handler) HealthCheckAppServer(ctx context.Context, publicAddr string, c
 	return nil
 }
 
-// handleForward forwards the request to the application service.
-func (h *Handler) handleForward(w http.ResponseWriter, r *http.Request, session *session) error {
+// handleHttp forwards the request to the application service or redirects
+// to the application directly.
+func (h *Handler) handleHttp(w http.ResponseWriter, r *http.Request, session *session) error {
+	redirected := false
+	session.tr.servers.Range(func(serverID, appServerInterface any) bool {
+		appServer, ok := appServerInterface.(types.AppServer)
+		if !ok {
+			h.log.Warnf("Failed to load AppServer, invalid type %T", appServerInterface)
+			return true
+		}
+
+		// If encounter an app server that is to be redirected to, stop iterating.
+		if redirectInsteadOfForward(appServer) {
+			redirected = true
+			http.Redirect(w, r, appServer.GetApp().GetURI(), http.StatusFound)
+			return true
+		}
+
+		return false
+	})
```

Comment:
This is effectively what is occurring here, but it's complicated by the way things work under the covers. The app servers are tested for matching to a public address, which can return more than one application. If you look at the current transport, this behaves in the same way:

```
	t.servers.Range(func(serverID, appServerInterface interface{}) bool {
		appServer, ok := appServerInterface.(types.AppServer)
		if !ok {
			t.c.log.Warnf("Failed to load AppServer, invalid type %T", appServerInterface)
			return true
		}

		var dialErr error
		conn, dialErr = dialAppServer(ctx, t.c.proxyClient, t.c.identity.RouteToApp.ClusterName, appServer)
		if dialErr != nil {
			if isReverseTunnelDownError(dialErr) {
				t.c.log.Warnf("Failed to connect to application server %q: %v.", serverID, dialErr)
				t.servers.Delete(serverID)
				// Only goes for the next server if the error returned is a
				// connection problem. Otherwise, stop iterating over the
				// servers and return the error.
				return true
			}
		}

		// "save" dial error to return as the function error.
		err = dialErr
		return false
	})
```

I think this logic, though somewhat unusual, maintains parity with the current behavior of the transport. However, I'll modify this to redirect after the range function.

Reply:
Oh snap, realized now that I inverted the bools for the loop -- will fix that too.